cirq-core/cirq/sim/mux.py

```diff
@@ -291,16 +293,50 @@ def final_density_matrix(
         return sparse_result.density_matrix_of()
     else:
         # noisy case: use DensityMatrixSimulator with dephasing
+        has_classical_control = circuit_like != measurement_transformers.defer_measurements(circuit_like)
+        handling_classical_control = ignore_measurement_results and has_classical_control
+
+        if handling_classical_control:
+            # case 1: classical control
+            noise_applied = circuit_like.with_noise(noise) if noise is not None else circuit_like
+            defered = measurement_transformers.defer_measurements(noise_applied)
+            dephased = measurement_transformers.dephase_measurements(defered)
+            program = dephased
+        elif ignore_measurement_results:
+            # case 2: no classical control, only terminal measurement
+            program = measurement_transformers.dephase_measurements(circuit_like)
+        else:
+            # case 3: no measurement
+            program = circuit_like
+
         density_result = density_matrix_simulator.DensityMatrixSimulator(
-            dtype=dtype, noise=noise, seed=seed
+            dtype=dtype, noise=None if handling_classical_control else noise, seed=seed
         ).simulate(
-            program=(
-                measurement_transformers.dephase_measurements(circuit_like)
-                if ignore_measurement_results
-                else circuit_like
-            ),
+            program,
             initial_state=initial_state,
             qubit_order=qubit_order,
             param_resolver=param_resolver,
         )
-        return density_result.final_density_matrix
+
+        res = density_result.final_density_matrix
+
+        if handling_classical_control:
+            num_qubits = qid_shape_protocol.num_qubits(circuit_like)
+            qid_shape = qid_shape_protocol.qid_shape(circuit_like)
+
+            # assuming that the ancella bits from the transformations are at the end
+            keep = list(range(num_qubits))
+
+            dephased_qid_shape = qid_shape_protocol.qid_shape(dephased)
+
+            tensor_form = np.reshape(res, dephased_qid_shape + dephased_qid_shape)
+
+            reduced_form = transformations.partial_trace(tensor_form, keep)
+
+            width = np.prod(qid_shape)
+
+            r = np.reshape(reduced_form, (width,width))
```

Comment:
Nit: either return directly here instead of assigning, or assign to `res` here so it'll return below.